# -*- Pyrex -*-
#
# ======================================================================
#
#                           Brad T. Aagaard
#                        U.S. Geological Survey
#
# {LicenseText}
#
# ======================================================================
#

#header{
#include "pylith/meshio/CellFilter.hh"
#include "pylith/meshio/CellFilterAvg.hh"
#include "pylith/meshio/DataWriter.hh"
#include "pylith/meshio/DataWriterVTK.hh"
#include "pylith/meshio/MeshIO.hh"
#include "pylith/meshio/MeshIOAscii.hh"
#include "pylith/meshio/MeshIOLagrit.hh"
#include "pylith/meshio/OutputManager.hh"
#include "pylith/meshio/OutputSolnSubset.hh"
#include "pylith/meshio/VertexFilter.hh"
#include "pylith/meshio/VertexFilterVecNorm.hh"

#if defined(ENABLE_CUBIT)
#include "pylith/meshio/MeshIOCubit.hh"
#endif

#include <strings.h> // USES strcasecmp()
#include <stdexcept>
#include <assert.h>
#include <Python.h>
#}header

# ----------------------------------------------------------------------
cdef extern from "Python.h":
  object PyCObject_FromVoidPtr(void*, void (*destruct)(void*))
  void* PyCObject_AsVoidPtr(object)

cdef void* ptrFromHandle(obj):
  """Extract pointer from PyCObject."""
  return PyCObject_AsVoidPtr(obj.handle)

cdef extern from "stdlib.h":
  ctypedef unsigned long size_t
  void* malloc(size_t size)
  void free(void* mem)

# ----------------------------------------------------------------------
cdef void MeshIO_destructor(void* obj):
  """
  Destroy MeshIO object.
  """
  # create shim for destructor
  #embed{ void MeshIO_destructor_cpp(void* pObj)
  pylith::meshio::MeshIO* io = (pylith::meshio::MeshIO*) pObj;
  delete io;
  #}embed
  MeshIO_destructor_cpp(obj)
  return

cdef class MeshIO:

  cdef void* thisptr # Pointer to C++ object
  cdef readonly object handle # PyCObject holding pointer to C++ object
  cdef readonly object name # Identifier for object base type

  def __init__(self):
    """
    Constructor.
    """
    self.handle = None
    self.thisptr = NULL
    self.name = "pylith_meshio_MeshIO"
    return


  def read(self, mesh):
    """
    Read mesh from file.
    """
    # create shim for method 'read'
    #embed{ void MeshIO_read(void* objVptr, void* meshVptr)
    try {
      assert(0 != objVptr);
      assert(0 != meshVptr);
      ALE::Obj<ALE::Mesh>* mesh = (ALE::Obj<ALE::Mesh>*) meshVptr;
      ((pylith::meshio::MeshIO*) objVptr)->read(mesh);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    if mesh.name != "pylith_topology_Mesh":
      raise TypeError, \
            "Argument must be extension module type 'Mesh'."
    MeshIO_read(self.thisptr, ptrFromHandle(mesh))
    return


  def write(self, mesh):
    """
    Read write to file.
    """
    # create shim for method 'write'
    #embed{ void MeshIO_write(void* objVptr, void* meshVptr)
    try {
      assert(0 != objVptr);
      ALE::Obj<ALE::Mesh>* mesh = (ALE::Obj<ALE::Mesh>*) meshVptr;
      ((pylith::meshio::MeshIO*) objVptr)->write(mesh);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    if not mesh.name == "pylith_topology_Mesh":
      raise TypeError, \
            "Argument must be extension module type 'Mesh'."
    MeshIO_write(self.thisptr, ptrFromHandle(mesh))
    return


  def _createHandle(self):
    """Wrap pointer to C++ object in PyCObject."""
    return PyCObject_FromVoidPtr(self.thisptr, MeshIO_destructor)


  property interpolate:
    def __set__(self, flag):
      """Set interpolate."""
      # create shim for method 'interpolate'
      #embed{ void MeshIO_interpolate_set(void* objVptr, int flag)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIO*) objVptr)->interpolate(flag);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIO_interpolate_set(self.thisptr, flag)

    def __get__(self):
      """Get interpolate."""
      # create shim for method 'interpolate'
      #embed{ int MeshIO_interpolate_get(void* objVptr)
      int result = 0;
      try {
        assert(0 != objVptr);
        result = ((pylith::meshio::MeshIO*) objVptr)->interpolate();
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIO_interpolate_get(self.thisptr)


  property debug:
    def __set__(self, flag):
      """Set debug."""
      # create shim for method 'debug'
      #embed{ void MeshIO_debug_set(void* objVptr, int flag)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIO*) objVptr)->debug(flag);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIO_debug_set(self.thisptr, flag)

    def __get__(self):
      """Get debug."""
      # create shim for method 'debug'
      #embed{ int MeshIO_debug_get(void* objVptr)
      int result = 0;
      try {
        assert(0 != objVptr);
        result = ((pylith::meshio::MeshIO*) objVptr)->debug();
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIO_debug_get(self.thisptr)


# ----------------------------------------------------------------------
cdef class MeshIOAscii(MeshIO):

  def __init__(self):
    """Constructor."""
    # create shim for constructor
    #embed{ void* MeshIOAscii_constructor()
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::MeshIOAscii);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #}embed

    MeshIO.__init__(self)
    self.thisptr = MeshIOAscii_constructor()
    self.handle = self._createHandle()
    return


  property filename:
    def __set__(self, name):
      """Set filename."""
      # create shim for method 'filename'
      #embed{ void MeshIOAscii_filename_set(void* objVptr, char* name)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIOAscii*) objVptr)->filename(name);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIOAscii_filename_set(self.thisptr, name)

    def __get__(self):
      """Get filename."""
      # create shim for method 'filename'
      #embed{ char* MeshIOAscii_filename_get(void* objVptr)
      char* result = 0;
      try {
        assert(0 != objVptr);
        result = (char*) ((pylith::meshio::MeshIOAscii*) objVptr)->filename();
        assert(0 != result);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIOAscii_filename_get(self.thisptr)


# ----------------------------------------------------------------------
cdef class MeshIOCubit(MeshIO):

  def __init__(self):
    """Constructor."""
    # create shim for constructor
    #embed{ void* MeshIOCubit_constructor()
    #if defined(ENABLE_CUBIT)
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::MeshIOCubit);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #else
    throw std::logic_error("Not configured to use CUBIT.");
    return 0;
    #endif
    #}embed

    MeshIO.__init__(self)
    self.thisptr = MeshIOCubit_constructor()
    self.handle = self._createHandle()
    return


  property filename:
    def __set__(self, name):
      """Set filename."""
      # create shim for method 'filename'
      #embed{ void MeshIOCubit_filename_set(void* objVptr, char* name)
      #if defined(ENABLE_CUBIT)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIOCubit*) objVptr)->filename(name);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #else
      throw std::logic_error("Not configured to use CUBIT.");
      #endif
      #}embed
      MeshIOCubit_filename_set(self.thisptr, name)

    def __get__(self):
      """Get filename."""
      # create shim for method 'filename'
      #embed{ char* MeshIOCubit_filename_get(void* objVptr)
      #if defined(ENABLE_CUBIT)
      char* result = 0;
      try {
        assert(0 != objVptr);
        result = (char*) ((pylith::meshio::MeshIOCubit*) objVptr)->filename();
        assert(0 != result);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #else
      throw std::logic_error("Not configured to use CUBIT.");
      return 0;
      #endif
      #}embed
      return MeshIOCubit_filename_get(self.thisptr)

# ----------------------------------------------------------------------
cdef class MeshIOLagrit(MeshIO):

  def __init__(self):
    """Constructor."""
    # create shim for constructor
    #embed{ void* MeshIOLagrit_constructor()
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::MeshIOLagrit);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #}embed

    MeshIO.__init__(self)
    self.thisptr = MeshIOLagrit_constructor()
    self.handle = self._createHandle()
    return


  property filenameGmv:
    def __set__(self, name):
      """Set filename."""
      # create shim for method 'filenameGmv'
      #embed{ void MeshIOLagrit_filenameGmv_set(void* objVptr, char* name)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIOLagrit*) objVptr)->filenameGmv(name);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIOLagrit_filenameGmv_set(self.thisptr, name)

    def __get__(self):
      """Get filename."""
      # create shim for method 'filenameGmv'
      #embed{ char* MeshIOLagrit_filenameGmv_get(void* objVptr)
      char* result = 0;
      try {
        assert(0 != objVptr);
        result = (char*) ((pylith::meshio::MeshIOLagrit*) objVptr)->filenameGmv();
        assert(0 != result);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIOLagrit_filenameGmv_get(self.thisptr)


  property filenamePset:
    def __set__(self, name):
      """Set filename."""
      # create shim for method 'filenamePset'
      #embed{ void MeshIOLagrit_filenamePset_set(void* objVptr, char* name)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIOLagrit*) objVptr)->filenamePset(name);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIOLagrit_filenamePset_set(self.thisptr, name)

    def __get__(self):
      """Get filename."""
      # create shim for method 'filenamePset'
      #embed{ char* MeshIOLagrit_filenamePset_get(void* objVptr)
      char* result = 0;
      try {
        assert(0 != objVptr);
        result = (char*) ((pylith::meshio::MeshIOLagrit*) objVptr)->filenamePset();
        assert(0 != result);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIOLagrit_filenamePset_get(self.thisptr)


  property flipEndian:
    def __set__(self, flag):
      """Set flip endian flag."""
      # create shim for method 'flipEndian'
      #embed{ void MeshIOLagrit_flipEndian_set(void* objVptr, int flag)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIOLagrit*) objVptr)->flipEndian(flag);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIOLagrit_flipEndian_set(self.thisptr, flag)

    def __get__(self):
      """Get flip endian flag."""
      # create shim for method 'flipEndian'
      #embed{ int MeshIOLagrit_flipEndian_get(void* objVptr)
      int result = 0;
      try {
        assert(0 != objVptr);
        result = ((pylith::meshio::MeshIOLagrit*) objVptr)->flipEndian();
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIOLagrit_flipEndian_get(self.thisptr)


  property ioInt32:
    def __set__(self, flag):
      """Set flip endian flag."""
      # create shim for method 'ioInt32'
      #embed{ void MeshIOLagrit_ioInt32_set(void* objVptr, int flag)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIOLagrit*) objVptr)->ioInt32(flag);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIOLagrit_ioInt32_set(self.thisptr, flag)

    def __get__(self):
      """Get flip endian flag."""
      # create shim for method 'ioInt32'
      #embed{ int MeshIOLagrit_ioInt32_get(void* objVptr)
      int result = 0;
      try {
        assert(0 != objVptr);
        result = ((pylith::meshio::MeshIOLagrit*) objVptr)->ioInt32();
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIOLagrit_ioInt32_get(self.thisptr)


  property isRecordHeader32Bit:
    def __set__(self, flag):
      """Set Fortran record header size flag."""
      # create shim for method 'isRecordHeader32Bit'
      #embed{ void MeshIOLagrit_isRecordHeader32Bit_set(void* objVptr, int flag)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::MeshIOLagrit*) objVptr)->isRecordHeader32Bit(flag);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      MeshIOLagrit_isRecordHeader32Bit_set(self.thisptr, flag)

    def __get__(self):
      """Get Fortran record header size flag."""
      # create shim for method 'isRecordHeader32Bit'
      #embed{ int MeshIOLagrit_isRecordHeader32Bit_get(void* objVptr)
      int result = 0;
      try {
        assert(0 != objVptr);
        result = ((pylith::meshio::MeshIOLagrit*) objVptr)->isRecordHeader32Bit();
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      return result;
      #}embed
      return MeshIOLagrit_isRecordHeader32Bit_get(self.thisptr)


# ----------------------------------------------------------------------
cdef void OutputManager_destructor(void* obj):
  """
  Destroy OutputManager object.
  """
  # create shim for destructor
  #embed{ void OutputManager_destructor_cpp(void* pObj)
  pylith::meshio::OutputManager* io = (pylith::meshio::OutputManager*) pObj;
  delete io;
  #}embed
  OutputManager_destructor_cpp(obj)
  return

cdef class OutputManager:

  cdef void* thisptr # Pointer to C++ object
  cdef readonly object handle # PyCObject holding pointer to C++ object
  cdef readonly object name # Identifier for object base type

  def __init__(self):
    """
    Constructor.
    """
    # create shim for constructor
    #embed{ void* OutputManager_constructor()
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::OutputManager);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #}embed

    self.name = "pylith_meshio_OutputManager"
    self.thisptr = OutputManager_constructor()
    self.handle = self._createHandle()
    return


  def open(self, mesh, cs, numTimeSteps, label, labelId):
    """
    Open output file.
    """
    # create shim for method 'open'
    #embed{ void OutputManager_open(void* objVptr, void* meshVptr, void* csVptr, int numTimeSteps, char* label, int labelId)
    try {
      assert(0 != objVptr);
      assert(0 != csVptr);
      ALE::Obj<ALE::Mesh>* mesh = (ALE::Obj<ALE::Mesh>*) meshVptr;
      spatialdata::geocoords::CoordSys* cs =
        (spatialdata::geocoords::CoordSys*) csVptr;
      ((pylith::meshio::OutputManager*) objVptr)->open(*mesh, cs, numTimeSteps, label, labelId);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    if mesh.name != "pylith_topology_Mesh":
      raise TypeError, \
            "Argument 'mesh' must be extension module type 'Mesh'."
    if not cs.name == "spatialdata_geocoords_CoordSys":
      raise TypeError, \
            "Argument 'cs' must be extension module type 'CoordSys'."
    if None == label or None == labelId:
      OutputManager_open(self.thisptr, ptrFromHandle(mesh),
                         ptrFromHandle(cs), numTimeSteps, NULL, 0)
    else:
      OutputManager_open(self.thisptr, ptrFromHandle(mesh),
                         ptrFromHandle(cs), numTimeSteps, label, labelId)
    return


  def close(self):
    """
    Close solution files.
    """
    # create shim for method 'close'
    #embed{ void OutputManager_close(void* objVptr)
    try {
      assert(0 != objVptr);
      ((pylith::meshio::OutputManager*) objVptr)->close();
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    OutputManager_close(self.thisptr)
    return


  def openTimeStep(self, t, mesh, cs, label, labelId):
    """
    Prepare for writing solution to file.
    """
    # create shim for method 'openTimeStep'
    #embed{ void OutputManager_openTimeStep(void* objVptr, double t, void* meshVptr, void* csVptr, char* label, int labelId)
    try {
      assert(0 != objVptr);
      assert(0 != meshVptr);
      assert(0 != csVptr);
      ALE::Obj<ALE::Mesh>* mesh = (ALE::Obj<ALE::Mesh>*) meshVptr;
      spatialdata::geocoords::CoordSys* cs =
        (spatialdata::geocoords::CoordSys*) csVptr;
      ((pylith::meshio::OutputManager*) objVptr)->openTimeStep(t, *mesh, cs,
                                                               label, labelId);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    if not mesh.name == "pylith_topology_Mesh":
      raise TypeError, \
            "Argument must be extension module type 'Mesh'."
    if not cs.name == "spatialdata_geocoords_CoordSys":
      raise TypeError, \
            "Argument must be extension module type 'CoordSys'."
    if None == label or None == labelId:
      OutputManager_openTimeStep(self.thisptr, t, 
                                 ptrFromHandle(mesh), ptrFromHandle(cs),
                                 NULL, 0)
    else:
      OutputManager_openTimeStep(self.thisptr, t, 
                                 ptrFromHandle(mesh), ptrFromHandle(cs),
                                 label, labelId)
    return


  def closeTimeStep(self):
    """
    Prepare for writing solution to file.
    """
    # create shim for method 'closeTimeStep'
    #embed{ void OutputManager_closeTimeStep(void* objVptr)
    try {
      assert(0 != objVptr);
      ((pylith::meshio::OutputManager*) objVptr)->closeTimeStep();
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    OutputManager_closeTimeStep(self.thisptr)
    return


  def appendVertexField(self, t, name, field, fieldType, mesh):
    """
    Append field over vertices to file.
    """
    # create shim for method 'appendVertexField'
    #embed{ void OutputManager_appendVertexField(void* objVptr, double t, char* name, void* fieldVptr, int fieldTypeInt, void* meshVptr)
    try {
      assert(0 != objVptr);
      assert(0 != fieldVptr);
      assert(0 != name);
      assert(0 != meshVptr);
      ALE::Obj<pylith::real_section_type>* field =
        (ALE::Obj<pylith::real_section_type>*) fieldVptr;
      ALE::Obj<ALE::Mesh>* mesh = (ALE::Obj<ALE::Mesh>*) meshVptr;
      const pylith::VectorFieldEnum fieldType =
        (pylith::VectorFieldEnum) fieldTypeInt;
      ((pylith::meshio::OutputManager*) objVptr)->appendVertexField(t, name,
                                                     *field, fieldType, *mesh);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    if not mesh.name == "pylith_topology_Mesh":
      raise TypeError, \
            "Argument 'mesh' must be extension module type 'Mesh'."
    cdef void* fieldVptr
    fieldVptr = PyCObject_AsVoidPtr(field)
    OutputManager_appendVertexField(self.thisptr, t, name, fieldVptr, fieldType, 
                                    ptrFromHandle(mesh))   
    return


  def appendCellField(self, t, name, field, fieldType, mesh, label, labelId):
    """
    Append field over cells to file.
    """
    # create shim for method 'appendCellField'
    #embed{ void OutputManager_appendCellField(void* objVptr, double t, char* name, void* fieldVptr, int fieldTypeInt, void* meshVptr, char* label, int labelId)
    try {
      assert(0 != objVptr);
      assert(0 != fieldVptr);
      assert(0 != name);
      assert(0 != meshVptr);
      ALE::Obj<pylith::real_section_type>* field =
        (ALE::Obj<pylith::real_section_type>*) fieldVptr;
      ALE::Obj<ALE::Mesh>* mesh = (ALE::Obj<ALE::Mesh>*) meshVptr;
      const pylith::VectorFieldEnum fieldType =
        (pylith::VectorFieldEnum) fieldTypeInt;
      ((pylith::meshio::OutputManager*) objVptr)->appendCellField(t, name,
                                                                  *field,
                                                                  fieldType,
                                                                  *mesh, label, labelId);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    #}embed
    if not mesh.name == "pylith_topology_Mesh":
      raise TypeError, \
            "Argument must be extension module type 'Mesh'."
    cdef void* fieldVptr
    fieldVptr = PyCObject_AsVoidPtr(field)
    if None == label or None == labelId:
      OutputManager_appendCellField(self.thisptr, t, name, fieldVptr, fieldType,
                                    ptrFromHandle(mesh), NULL, 0)
    else:
      OutputManager_appendCellField(self.thisptr, t, name, fieldVptr, fieldType,
                                    ptrFromHandle(mesh), label, labelId)
    return


  def _createHandle(self):
    """Wrap pointer to C++ object in PyCObject."""
    return PyCObject_FromVoidPtr(self.thisptr, OutputManager_destructor)


  property coordsys:
    def __set__(self, cs):
      """Set coordinate system."""
      # create shim for method 'coordsys'
      #embed{ void OutputManager_coordsys_set(void* objVptr, void* csVptr)
      try {
        assert(0 != objVptr);
        assert(0 != csVptr);
        spatialdata::geocoords::CoordSys* cs =
          (spatialdata::geocoords::CoordSys*) csVptr;
        ((pylith::meshio::OutputManager*) objVptr)->coordsys(cs);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      OutputManager_coordsys_set(self.thisptr, ptrFromHandle(cs))


  property writer:
    def __set__(self, writer):
      """
      Writer for finite-element data.
      """
      # create shim for method 'writer'
      #embed{ void OutputManager_writer_set(void* objVptr, void* writerVptr)
      try {
        assert(0 != objVptr);
        assert(0 != writerVptr);
        pylith::meshio::DataWriter* writer = 
          (pylith::meshio::DataWriter*) writerVptr;
        ((pylith::meshio::OutputManager*) objVptr)->writer(writer);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      OutputManager_writer_set(self.thisptr, ptrFromHandle(writer))


  property vertexFilter:
    def __set__(self, filter):
      """
      Set vertex filter.
      """
      # create shim for method 'vertexFilter'
      #embed{ void OutputManager_vertexFilter_set(void* objVptr, void* filterVptr)
      try {
        assert(0 != objVptr);
        pylith::meshio::VertexFilter* filter = 
          (pylith::meshio::VertexFilter*) filterVptr;
        ((pylith::meshio::OutputManager*) objVptr)->vertexFilter(filter);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      if filter != None:
        OutputManager_vertexFilter_set(self.thisptr, ptrFromHandle(filter))
      else:
        OutputManager_vertexFilter_set(self.thisptr, NULL)


  property cellFilter:
    def __set__(self, filter):
      """
      Set cell filter.
      """
      # create shim for method 'cellFilter'
      #embed{ void OutputManager_cellFilter_set(void* objVptr, void* filterVptr)
      try {
        assert(0 != objVptr);
        pylith::meshio::CellFilter* filter = 
          (pylith::meshio::CellFilter*) filterVptr;
        ((pylith::meshio::OutputManager*) objVptr)->cellFilter(filter);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      if filter != None:
        OutputManager_cellFilter_set(self.thisptr, ptrFromHandle(filter))
      else:
        OutputManager_cellFilter_set(self.thisptr, NULL)


# ----------------------------------------------------------------------
cdef class OutputSolnSubset(OutputManager):

  def __init__(self):
    """
    Constructor.
    """
    # create shim for constructor
    #embed{ void* OutputSolnSubset_constructor()
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::OutputSolnSubset);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #}embed

    self.name = "pylith_meshio_OutputSolnSubset"
    self.thisptr = OutputSolnSubset_constructor()
    self.handle = self._createHandle()
    return


  def mesh(self, submesh, mesh):
    """
    Get mesh associated with subdomain.
    """
    # create shim for method 'subdomainMesh'
    #embed{ void OutputSolnSubset_subdomainMesh(void* objVptr, void* submeshVptr, void* meshVptr)
    try {
      assert(0 != objVptr);
      assert(0 != submeshVptr);
      assert(0 != meshVptr);
      ALE::Obj<ALE::Mesh>* submesh = (ALE::Obj<ALE::Mesh>*) submeshVptr;
      ALE::Obj<ALE::Mesh>* mesh = (ALE::Obj<ALE::Mesh>*) meshVptr;
      *submesh =
        ((pylith::meshio::OutputSolnSubset*) objVptr)->subdomainMesh(*mesh);
      } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch      
    #}embed

    if not submesh.name == "pylith_topology_Mesh":
      raise TypeError, \
            "Argument 'submesh' must be extension module type " \
            "'pylith::topology::Mesh'."
    if not mesh.name == "pylith_topology_Mesh":
      raise TypeError, \
            "Argument 'mesh' must be extension module type " \
            "'pylith::topology::Mesh'."
    OutputSolnSubset_subdomainMesh(self.thisptr,
                                   ptrFromHandle(submesh), ptrFromHandle(mesh))
    return
    

  property label:
    def __set__(self, value):
      """
      Set label of subdomain.
      """
      # create shim for method 'label'
      #embed{ void OutputSolnSubset_label_set(void* objVptr, char* value)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::OutputSolnSubset*) objVptr)->label(value);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      OutputSolnSubset_label_set(self.thisptr, value)


# ----------------------------------------------------------------------
cdef void DataWriter_destructor(void* obj):
  """
  Destroy DataWriter object.
  """
  # create shim for destructor
  #embed{ void DataWriter_destructor_cpp(void* pObj)
  pylith::meshio::DataWriter* io = (pylith::meshio::DataWriter*) pObj;
  delete io;
  #}embed
  DataWriter_destructor_cpp(obj)
  return

cdef class DataWriter:

  cdef void* thisptr # Pointer to C++ object
  cdef readonly object handle # PyCObject holding pointer to C++ object
  cdef readonly object name # Identifier for object base type

  def __init__(self):
    """
    Constructor.
    """
    self.handle = None
    self.thisptr = NULL
    self.name = "pylith_meshio_DataWriter"
    return


  def _createHandle(self):
    """Wrap pointer to C++ object in PyCObject."""
    return PyCObject_FromVoidPtr(self.thisptr, DataWriter_destructor)


# ----------------------------------------------------------------------
cdef class DataWriterVTK(DataWriter):

  def __init__(self):
    """Constructor."""
    # create shim for constructor
    #embed{ void* DataWriterVTK_constructor()
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::DataWriterVTK);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #}embed

    DataWriter.__init__(self)
    self.thisptr = DataWriterVTK_constructor()
    self.handle = self._createHandle()
    return


  property filename:
    def __set__(self, name):
      """Set filename."""
      # create shim for method 'filename'
      #embed{ void DataWriterVTK_filename_set(void* objVptr, char* name)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::DataWriterVTK*) objVptr)->filename(name);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      DataWriterVTK_filename_set(self.thisptr, name)


  property timeFormat:
    def __set__(self, format):
      """Set timeFormat."""
      # create shim for method 'timeFormat'
      #embed{ void DataWriterVTK_timeFormat_set(void* objVptr, char* format)
      try {
        assert(0 != objVptr);
        ((pylith::meshio::DataWriterVTK*) objVptr)->timeFormat(format);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      DataWriterVTK_timeFormat_set(self.thisptr, format)


# ----------------------------------------------------------------------
cdef void CellFilter_destructor(void* obj):
  """
  Destroy CellFilter object.
  """
  # create shim for destructor
  #embed{ void CellFilter_destructor_cpp(void* pObj)
  pylith::meshio::CellFilter* f = (pylith::meshio::CellFilter*) pObj;
  delete f;
  #}embed
  CellFilter_destructor_cpp(obj)
  return

cdef class CellFilter:

  cdef void* thisptr # Pointer to C++ object
  cdef readonly object handle # PyCObject holding pointer to C++ object
  cdef readonly object name # Identifier for object base type

  def __init__(self):
    """
    Constructor.
    """
    self.handle = None
    self.thisptr = NULL
    self.name = "pylith_meshio_CellFilter"
    return


  property quadrature:
    def __set__(self, quadrature):
      """
      Set quadrature.
      """
      # create shim for method 'quadrature'
      #embed{ void CellFilter_quadrature_set(void* objVptr, void* qVptr)
      try {
        assert(0 != objVptr);
        assert(0 != qVptr);
        pylith::feassemble::Quadrature* quadrature =
          (pylith::feassemble::Quadrature*) qVptr;
        ((pylith::meshio::CellFilter*) objVptr)->quadrature(quadrature);
      } catch (const std::exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.what()));
      } catch (const ALE::Exception& err) {
        PyErr_SetString(PyExc_RuntimeError,
                        const_cast<char*>(err.msg().c_str()));
      } catch (...) {
        PyErr_SetString(PyExc_RuntimeError,
                        "Caught unknown C++ exception.");
      } // try/catch
      #}embed
      if not quadrature.name == "pylith_feassemble_Quadrature":
        raise TypeError, \
              "Argument must be extension module type 'Quadrature'."
      CellFilter_quadrature_set(self.thisptr, ptrFromHandle(quadrature))


  def _createHandle(self):
    """Wrap pointer to C++ object in PyCObject."""
    return PyCObject_FromVoidPtr(self.thisptr, CellFilter_destructor)


# ----------------------------------------------------------------------
cdef class CellFilterAvg(CellFilter):

  def __init__(self):
    """Constructor."""
    # create shim for constructor
    #embed{ void* CellFilterAvg_constructor()
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::CellFilterAvg);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #}embed

    CellFilter.__init__(self)
    self.thisptr = CellFilterAvg_constructor()
    self.handle = self._createHandle()
    return


# ----------------------------------------------------------------------
cdef void VertexFilter_destructor(void* obj):
  """
  Destroy VertexFilter object.
  """
  # create shim for destructor
  #embed{ void VertexFilter_destructor_cpp(void* pObj)
  pylith::meshio::VertexFilter* f = (pylith::meshio::VertexFilter*) pObj;
  delete f;
  #}embed
  VertexFilter_destructor_cpp(obj)
  return

cdef class VertexFilter:

  cdef void* thisptr # Pointer to C++ object
  cdef readonly object handle # PyCObject holding pointer to C++ object
  cdef readonly object name # Identifier for object base type

  def __init__(self):
    """
    Constructor.
    """
    self.handle = None
    self.thisptr = NULL
    self.name = "pylith_meshio_VertexFilter"
    return


  def _createHandle(self):
    """Wrap pointer to C++ object in PyCObject."""
    return PyCObject_FromVoidPtr(self.thisptr, VertexFilter_destructor)


# ----------------------------------------------------------------------
cdef class VertexFilterVecNorm(VertexFilter):

  def __init__(self):
    """Constructor."""
    # create shim for constructor
    #embed{ void* VertexFilterVecNorm_constructor()
    void* result = 0;
    try {
      result = (void*)(new pylith::meshio::VertexFilterVecNorm);
      assert(0 != result);
    } catch (const std::exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.what()));
    } catch (const ALE::Exception& err) {
      PyErr_SetString(PyExc_RuntimeError,
                      const_cast<char*>(err.msg().c_str()));
    } catch (...) {
      PyErr_SetString(PyExc_RuntimeError,
                      "Caught unknown C++ exception.");
    } // try/catch
    return result;
    #}embed

    VertexFilter.__init__(self)
    self.thisptr = VertexFilterVecNorm_constructor()
    self.handle = self._createHandle()
    return


# End of file
